### PR TITLE
Add RAGEval banckend with `MTEB` banckmark and `ragas` framework

### PR DESCRIPTION
**Pull request by Yunnglin**
## MTEB benchmark
- support `MTEB` and `CMTEB` benchmark
- support auto download model and datasets from modelscope
- Support one-stage evaluation for embedding models, including classification, clustering, pair-classification, reranking, retrieval, and sts.
- support two-stage evaluation for cross-encoder (reranker) models: two-stage run with the second stage reading results saved from the first stage.

example:
```python
# Copyright (c) Alibaba, Inc. and its affiliates.

"""
1. Installation
EvalScope: pip install mteb

2. Run eval task
"""
from evalscope.run import run_task
from evalscope.utils.logger import get_logger
import torch

logger = get_logger()


def run_eval():

    # Prepare the config
    model_name1 = "Jerry0/m3e-base"

    # model_name = "OpenBMB//embedding/MiniCPM-Embedding"
    # model_name = "Xorbits/bge-reranker-base"
    # model_name2 = "OpenBMB/MiniCPM-Reranker"
    model_name2 = "jinaai/jina-reranker-v2-base-multilingual"

    one_stage_task_cfg = {
        "eval_backend": "RAGEval",
        "eval_config": {
            "tool": "MTEB",
            "model": [
                {
                    "model_name_or_path": model_name1,
                    "pooling_mode": None,  # load from model config
                    "max_seq_length": 512,
                    "prompt": "为这个问题生成一个检索用的表示",
                    "model_kwargs": {"torch_dtype": "auto"},
                    "encode_kwargs": {
                        "batch_size": 32,
                    },
                }
            ],
            "eval": {
                "tasks": ["T2Retrieval"],
                "verbosity": 2,
                "output_folder": "outputs",
                "overwrite_results": True,
                "limits": 100,
            },
        },
    }

    two_stage_task_cfg = {
        "eval_backend": "RAGEval",
        "eval_config": {
            "tool": "MTEB",
            "model": [
                {
                    "model_name_or_path": model_name1,
                    "is_cross_encoder": False,
                    "max_seq_length": 512,
                    "prompt": "",
                    "model_kwargs": {"torch_dtype": "auto"},
                    "encode_kwargs": {
                        "batch_size": 32,
                    },
                },
                {
                    "model_name_or_path": model_name2,
                    "is_cross_encoder": True,
                    "max_seq_length": 512,
                    "prompt": "请根据问题生成一个检索用的表示",
                    "model_kwargs": {"torch_dtype": torch.float16},
                    "encode_kwargs": {
                        "batch_size": 32,
                    },
                },
            ],
            "eval": {
                "tasks": ["T2Retrieval"],
                "verbosity": 2,
                "output_folder": "outputs",
                "overwrite_results": True,
                "limits": 100,
            },
        },
    }


    # Run task
    run_task(task_cfg=two_stage_task_cfg)


if __name__ == "__main__":
    run_eval()

```

## ragas framework

example:

```python
from evalscope.run import run_task
from evalscope.utils.logger import get_logger

logger = get_logger()


def run_eval():
    generate_testset_task_cfg = {
        "eval_backend": "RAGEval",
        "eval_config": {
            "tool": "RAGAS",
            "testset_generation": {
                "docs": ["README.md"],
                "test_size": 5,
                "output_file": "outputs/testset.json",
                "distribution": {"simple": 0.5, "multi_context": 0.4, "reasoning": 0.1},
                # The generator_llm is the component that generates the questions, and evolves the question to make it more relevant.
                # The critic_llm is the component that filters the questions and nodes based on the question and node relevance.
                "generator_llm": {
                    "model_name_or_path": "qwen/Qwen2-7B-Instruct",
                    "template_type": "qwen",
                },
                "critic_llm": {
                    "model_name_or_path": "QwenCollection/Ragas-critic-llm-Qwen1.5-GPTQ",
                    "template_type": "qwen",
                },
                "embeddings": {
                    "model_name_or_path": "AI-ModelScope/m3e-base",
                },
            }
        },
    }
    
    eval_task_cfg = {
        "eval_backend": "RAGEval",
        "eval_config": {
            "tool": "RAGAS",
            "eval": {
                "testset_file": "outputs/testset.json",
                "critic_llm": {
                    "model_name_or_path": "qwen/Qwen2-7B-Instruct",
                    "template_type": "qwen",
                },
                "embeddings": {
                    "model_name_or_path": "AI-ModelScope/m3e-base",
                },
                "metrics": [
                    "faithfulness",
                    "answer_relevancy",
                    "context_precision",
                    "answer_correctness",
                ],
            },
        },
    }

    # Run task
    run_task(task_cfg=eval_task_cfg) # or run_task(task_cfg=generate_testset_task_cfg)


if __name__ == "__main__":
    run_eval()

```

### CODE REVIEW

File: evalscope/backend/rag_eval/backend_manager.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        eval_args = EvalArguments(**self.eval_args).to_dict()
+
+        if len(model_args_list) == 1:
+            one_stage_eval(model_args_list[0], eval_args)
+        else:  # len(model_args_list) == 2
+            two_stage_eval(model_args_list[0], model_args_list[1], eval_args)
+
+    def run(self, *args, **kwargs):
+        tool = self.config_d.pop("tool")
+        if tool.lower() == "MTEB":
```
Comment (yingdachen) [inline evalscope/backend/rag_eval/backend_manager.py]: can this check ever pass?

File: evalscope/backend/rag_eval/ragas/task_template.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import os
```
Comment (yingdachen) [inline evalscope/backend/rag_eval/ragas/task_template.py:1]: [evalscope/backend/rag_eval/ragas/arguments.py] 这个文件是空的？

File: evalscope/backend/rag_eval/ragas/task_template.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+    # generator with openai models
+    generator_llm = LLM(model_name_or_path="qwen/Qwen2-7B-Instruct", template_type="qwen")
+    critic_llm = LLM(
+        model_name_or_path="QwenCollection/Ragas-critic-llm-Qwen1.5-GPTQ",
+        template_type="qwen",
+    )
+    embeddings = EmbeddingModel.from_pretrained(
+        model_name_or_path="Jerry0/m3e-base", is_cross_encoder=False
```
Comment (yingdachen) [inline evalscope/backend/rag_eval/ragas/task_template.py]: 用/ai-modelscope/m3e-base的正式一点？

File: evalscope/backend/rag_eval/ragas/task_template.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        if evolution.node_filter is None:
+            evolution.node_filter = node_filter
+
+        if isinstance(evolution, ComplexEvolution):
+            if evolution.evolution_filter is None:
+                evolution.evolution_filter = evolution_filter
+
+    generator = TestsetGenerator.from_langchain(generator_llm, critic_llm, embeddings)
+
+    # runconfig = RunConfig(timeout=30, max_retries=1, max_wait=30, max_workers=1)
```
Comment (yingdachen) [inline evalscope/backend/rag_eval/ragas/task_template.py]: remove if not needed

File: evalscope/backend/rag_eval/ragas/task_template.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+def rag_eval(
+    args,
+) -> None:
+    from datasets import Dataset
+    from ragas import evaluate, RunConfig
+    from ragas.metrics import faithfulness, answer_correctness
+    from evalscope.backend.rag_eval import EmbeddingModel, LLM
+
+    llm = LLM(model_name_or_path="qwen/Qwen2-7B-Instruct", template_type="qwen")
+    embedding = EmbeddingModel.from_pretrained(model_name_or_path="Jerry0/m3e-base", is_cross_encoder=False)
```
Comment (yingdachen) [inline evalscope/backend/rag_eval/ragas/task_template.py]: 同上

File: evalscope/backend/rag_eval/backend_manager.py (hunk 1 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/backend_manager.py]: fixed

File: evalscope/backend/rag_eval/ragas/task_template.py (hunk 2 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/ragas/task_template.py:1]: updated

File: evalscope/backend/rag_eval/ragas/task_template.py (hunk 3 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/ragas/task_template.py]: done

File: evalscope/backend/rag_eval/ragas/task_template.py (hunk 4 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/ragas/task_template.py]: done

File: evalscope/backend/rag_eval/ragas/task_template.py (hunk 5 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/ragas/task_template.py]: done

File: evalscope/backend/opencompass/__init__.py [hunk 6]
```diff
@@ -1,3 +1,3 @@
 # Copyright (c) Alibaba, Inc. and its affiliates.
 
-from evalscope.backend.opencompass.backend_manager import OpenCompassBackendManager
+from .backend_manager import OpenCompassBackendManager
```
Comment (wangxingjun778) [inline evalscope/backend/opencompass/__init__.py]: 使用绝对路径import，下同。

File: requirements/rag.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ragas
+mteb==0.14.16
```
Comment (wangxingjun778) [inline requirements/rag.txt]: 这里可否设置为 >= ，而非==

File: evalscope/backend/rag_eval/backend_manager.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            TestsetGenerationArguments,
+            EvaluationArguments,
+        )
+
+        if self.testset_args is not None:
+            testset_generation(TestsetGenerationArguments(**self.testset_args))
+        if self.eval_args is not None:
+            rag_eval(EvaluationArguments(**self.eval_args))
+
+    def run(self, *args, **kwargs):
```
Comment (wangxingjun778) [inline evalscope/backend/rag_eval/backend_manager.py:55]: 这里不同的tool，如`mteb`和`ragas`，建议在tests模块中增加UT，确保核心链路能够被覆盖到。

File: evalscope/backend/opencompass/__init__.py (hunk 6 above)
Comment (Yunnglin) [reply evalscope/backend/opencompass/__init__.py]: fixed

File: requirements/rag.txt (hunk 7 above)
Comment (Yunnglin) [reply requirements/rag.txt]: fixed

File: evalscope/backend/rag_eval/backend_manager.py (hunk 8 above)
Comment (Yunnglin) [reply evalscope/backend/rag_eval/backend_manager.py:55]: added

**Pull request merged**